### DISCUSSION TITLE
전략패턴의 장단점을 경험과 함께 들려주세요

### DISCUSSION THREAD

**Discussion by 3Juhwan** [2024-woowacourse-study/level-interview#15] [OOP]
.

**Comment by gyungchan-jo** [comment]
우테코 레벨1 사다리 게임 미션에서 사다리를 생성하는 로직을 구현하기 위해 전략패턴을 적용했었습니다.

이를 통해 얻은 장점은 테스트를 쉽게 구현할 수 있었습니다.
랜덤으로 생성한 사다리를 테스트하는 것은 어렵기 때문에, 랜덤으로 사다리를 생성하는 객체와 지정한 값으로 사다리를 생성하는 객체를 따로 구현하여 테스트를 쉽게 구현할 수 있었습니다.

단점으로는 관리해야할 클래스가 증가하여 코드의 복잡도가 증가한다는 단점이 있었습니다.

**Comment by le2sky** [comment]
객체의 행위를 다양하게 관리하기 위해서 사용할 수 있습니다.
객체의 행위를 다른 것으로 바꾸고 싶은 경우, 직접 코드를 수정하는 것이 아닌
캡슐화된 알고리즘을 교체하도록 합니다.

```java
package racinggame.domain;

import java.util.Objects;

public class Car {

    ....
    public void move(MoveCondition moveCondition) {
        if (moveCondition.isMovable()) {
            position = position.increase();
        }
    }
    ....
}

```

자동차 경주 미션을 진행하면서, 전략 패턴을 사용한 경험이 있습니다.
자동차는 랜덤으로 값을 하나 뽑아서 특정 값 이상인 경우 이동할 수 있습니다.
이때, 자동차의 이동 조건을 추상화했습니다. 
테스트할 때는 다른 이동 조건을 자동차에게 주입하여, 테스트하기 쉽도록 만든 경험이 있습니다.
이때, 코드의 복잡도가 증가하는 것을 느꼈습니다.
정말 필요한 경우에만 사용해야겠다고 생각했습니다.
